h2integrate/control/control_strategies/pyomo_controllers.py

```diff
@@ -0,0 +1,757 @@
+    resource_name: str = field()
+    resource_storage_units: str = field()
+    tech_name: str = field()
+
+
+def dummy_function():
+    return None
+
+
+class PyomoControllerBaseClass(ControllerBaseClass):
+    def dummy_method(self, in1, in2):
+        return None
+
+    def setup(self):
+        # get technology group name
+        # TODO: Make this more general, right now it might go astray if for example "battery" is
+        # used twice in an OpenMDAO subsystem pathname
+        self.tech_group_name = self.pathname.split(".")
+
+        # create inputs for all pyomo object creation functions from all connected technologies
+        self.dispatch_connections = self.options["plant_config"]["tech_to_dispatch_connections"]
+        for connection in self.dispatch_connections:
+            source_tech, intended_dispatch_tech = connection
+            if any(intended_dispatch_tech in name for name in self.tech_group_name):
+                if source_tech == intended_dispatch_tech:
+                    # When getting rules for the same tech, the tech name is not used in order to
+                    # allow for automatic connections rather than complicating the h2i model set up
+                    self.add_discrete_input("dispatch_block_rule_function", val=self.dummy_method)
+                else:
+                    self.add_discrete_input(
+                        f"{'dispatch_block_rule_function'}_{source_tech}", val=self.dummy_method
+                    )
+            else:
+                continue
+
+        # create output for the pyomo control model
+        self.add_discrete_output(
+            "pyomo_dispatch_solver",
+            val=dummy_function,
+            desc="callable: fully formed pyomo model and execution logic to be run \
+                by owning technologies performance model",
+        )
+
+    def compute(self, inputs, outputs, discrete_inputs, discrete_outputs):
+        discrete_outputs["pyomo_dispatch_solver"] = self.pyomo_setup(discrete_inputs)
+
+    def pyomo_setup(self, discrete_inputs):
+        # initialize the pyomo model
+        self.pyomo_model = pyomo.ConcreteModel()
+
+        index_set = pyomo.Set(initialize=range(self.config.n_control_window))
+
+        # run each pyomo rule set up function for each technology
+        for connection in self.dispatch_connections:
+            source_tech, intended_dispatch_tech = connection
+            if any(intended_dispatch_tech in name for name in self.tech_group_name):
+                if source_tech == intended_dispatch_tech:
+                    dispatch_block_rule_function = discrete_inputs["dispatch_block_rule_function"]
+                else:
+                    dispatch_block_rule_function = discrete_inputs[
+                        f"{'dispatch_block_rule_function'}_{source_tech}"
+                    ]
+                blocks = pyomo.Block(index_set, rule=dispatch_block_rule_function)
+                setattr(self.pyomo_model, source_tech, blocks)
```

Comment:
Logic like this is in a few spots and it's not immediately clear what the outcome of it is (read: took me a few tries to understand this). Could you add more exhaustive comments explaining what's going on here, and maybe even what it looks like in the `plant_config.yaml` and how that gets used here?

Reply:
I've added clarifying comments